src/Microsoft.AspNet.OData.Shared/Builder/DerivedTypeConstraintSet.cs

```diff
@@ -0,0 +1,94 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.OData.Edm.Csdl;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    /// <summary>
+    /// Set of derived type constraints. 
+    /// </summary>
+    public class DerivedTypeConstraintSet : HashSet<Type>
+    {
+        /// <summary>
+        /// Initializes a DerivedTypeConstraintSet instance without setting a base type.
+        /// </summary>
+        public DerivedTypeConstraintSet()
+        {
+            Location = EdmVocabularyAnnotationSerializationLocation.Inline;
+        }
+
+        /// <summary>
+        /// Initializes a DerivedTypeContraintSet instance with a base type specified.
+        /// </summary>
+        /// <param name="baseType">CLR type of the base class.</param>
+        public DerivedTypeConstraintSet(Type baseType)
+            : base()
+        {
+            ClrBaseType = baseType;
+        }
+
+        /// <summary>
+        /// CLR Type of the base class. 
+        /// </summary>
+        public Type ClrBaseType { get; set; }
+
+        /// <summary>
+        /// Location of the derived type constraint annotation. By default, it will be inline.
+        /// </summary>
+        public EdmVocabularyAnnotationSerializationLocation Location { get; set; }
+
+        /// <summary>
+        /// Add TDerivedType to the set of derived type constraints.
+        /// </summary>
+        /// <typeparam name="TDerivedType">Derived Type.</typeparam>
+        /// <returns>Updated DerivedTypeConstraint set.</returns>
+        public DerivedTypeConstraintSet AddConstraint<TDerivedType>()
+        {
+            ValidateAndAddSingleConstraint(typeof(TDerivedType));
+            return this;
+        }
+
+        /// <summary>
+        /// Add the derived type constraints to the set. 
+        /// </summary>
+        /// <param name="derivedTypes">Derived types to be added to the set.</param>
+        public void AddConstraints(IEnumerable<Type> derivedTypes)
+        {
+            if (!derivedTypes.Any())
+            {
+                ValidateAndAddSingleConstraint(ClrBaseType);
+                return;
+            }
+
+            foreach (Type type in derivedTypes)
+            {
+                ValidateAndAddSingleConstraint(type);
+            }
+        }
+
+        private void ValidateAndAddSingleConstraint(Type typeToAdd)
+        {
+            if (ClrBaseType == null)
+            {
+                throw Error.InvalidOperation(SRResources.NoClrTypeSpecified);
+            }
+            else if (!ClrBaseType.IsAssignableFrom(typeToAdd) && typeToAdd != ClrBaseType)
+            {
+                throw Error.InvalidOperation(SRResources.NotADerivedType, typeToAdd.Name, ClrBaseType.Name);
+            }
+
+            if (!this.Add(typeToAdd))
+            {
+                throw Error.InvalidOperation(SRResources.ConstraintAlreadyExists, typeToAdd.Name);
+            }
+        }
+    }
```

Comment:
I'd like to copy/paste my assumption:
I don't want to use my name, but, for the "Add" logic, we can make it simple as blow

    /// <summary>
    /// 
    /// </summary>
    public class DerivedTypeConstraintConfiguration
    {
        private ISet<Type> _derivedTypes;

        /// <summary>
        /// 
        /// </summary>
        /// <param name="baseType"></param>
        public DerivedTypeConstraintConfiguration(Type baseType)
        {
            ClrBaseType = baseType;
            _derivedTypes = new HashSet<Type>();
            Location = EdmVocabularyAnnotationSerializationLocation.OutOfLine;
        }

        /// <summary>
        /// 
        /// </summary>
        public ISet<Type> DerivedTypes { get { return _derivedTypes; } }

        /// <summary>
        /// 
        /// </summary>
        public Type ClrBaseType { get; private set; }

        /// <summary>
        /// 
        /// </summary>
        /// <typeparam name="TDerived"></typeparam>
        public void Add<TDerived>()
        {
            Add(typeof(TDerived));
        }

        /// <summary>
        /// 
        /// </summary>
        public EdmVocabularyAnnotationSerializationLocation Location { get; set; }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="derivedType"></param>
        public void Add(Type derivedType)
        {
            if (derivedType == ClrBaseType)
            {
                return;
            }

            if (_derivedTypes.Contains(derivedType))
            {
                return;
            }

            if (!ClrBaseType.IsAssignableFrom(derivedType))
            {
                throw new Exception("blabla");
            }

            _derivedTypes.Add(derivedType);
        }
    }

Reply:
The main difference that I see is that you don't support adding enumerables. I think it makes the implementation a lot cleaner if you can add enumerables to the set. 

---
In reply to: [394509903](https://github.com/OData/WebApi/pull/1977#discussion_r394509903) [](ancestors = 394509903)